16364 fullscreen edit

ulferts opened the pull request:
Supersedes #3690 due to git history clean-up. The original PR had the following description:
# 

As the previous pull request contained breaking changes, a new one had to be opened.
The work here is related to the link below.

See #3646 and #3661 for detailed information about the intentions.
Both PRs were merged into this one.

https://community.openproject.org/work_packages/16364/activity
# 
## Known bugs to be fixed before merge
- [x] No conflicting modification warning upon saving. The fields are simply not saved. There is a js error shown in the console
- [x] Assignee and responsible not updated after successful editing (reload required)
- [x] Trying to edit "Estimated time" leads to a non terminating digest cycle
- [x] linked resources e.g. responsible, assignee, version, category can not be set to none.
- [x] Fields inherited from children not disabled/uneditable: Priority, Estimated time (https://github.com/opf/openproject/pull/3728)
- [x] Wrong placeholder shown for description
- [x] Editing a field without clicking into it (e.g. by using tab) will not register the field for submission.
- [x] Type switching does not update the fields available (e.g. custom fields)
  - https://community.openproject.org/work_packages/21302
## Known bugs that do not necessarily have to be fixed before merge
- [x] Pressing "ESC" while being in the full edit mode, will lead to the one field leaving the edit mode.
- [x] When the work package has a child and the type is switched to milestone (which can not have children), the date field disappears.
  - Not solved perfectly.  For now, the front end prevents fields, that are not editable (e.g. because they are inherited as the date is in that case) to be switched to edit mode. 
## Other todos before merging
- [x] fix bugs on configurations before merging

:warning: **Merge https://github.com/finnlabs/openproject-costs/pull/178 together with this PR** :warning:

furinvader committed:
Implement work package editing in fullscreen view

furinvader committed:
Upgrade protractor to v2.4.0

furinvader committed:
Implement tests for WP fullscreen editing

furinvader committed:
Implement WP editing in list  overview

furinvader committed:
Add work package edit actions toolbar (save & cancel)

Saving the work package does not work as of yet.

furinvader committed:
Disable WP edit buttons for users without update permissions

furinvader committed:
Update lodash to 2.4.2

furinvader committed:
Refactor EditableFieldsService and EditPaneController

Move isActiveField and getPendingFormChanges to EditableFieldsState service.

furinvader committed:
Implement WP-edit save button

furinvader committed:
Make user input stay when switching between views

furinvader committed:
Reset user input when cancel button is used in edit mode

furinvader committed:
Change appearance of edit mode action buttons

furinvader committed:
Implement cancel button for fullscreen edit

Input values are set to their initial values when one restarts editing.

furinvader committed:
Change edit action toolbar appearance

furinvader committed:
Disable some ui elements when in edit mode

furinvader committed:
Ask user for confirmation when changing states

[ci skip]

furinvader committed:
Replace UI Select with default HTML elements

[ci skip]

furinvader committed:
Simplify component handling

[ci skip]

furinvader committed:
Refactor HALAPIResource service

furinvader committed:
Refactor inplace-edit directory structure

[ci skip]

furinvader committed:
Refactor component initiation of inplace-edit

Well, I actually moved a line in the HALAPIResource service,
too...

furinvader committed:
Reference field names via newly introduced field object

furinvader committed:
Allow tests/specs to be located under components/

This makes it possible to have tests in the same
 location as the components they belong to.

furinvader committed:
Reduce fieldController redundancies

[ci skip]

furinvader committed:
Fix error that prevented protractor from starting

[ci skip]

furinvader committed:
Keep inplace edit values between views

... while also keeping empty values

[ci skip]

furinvader committed:
Set fields to edit state when creating a new work pacakge

[ci skip]

furinvader committed:
Make inplaceEdit.from method an optional getter

Provide a length property to indicate the number of fields
associated with that form.
The changes also enhance testability.

furinvader committed:
Add missing dependency to WorkPackageFieldService

$filter was missing due to faulty resolved rebasing.

ulferts committed:
replace ng-include by custom dom node

This fixes previewing newly to be added comments. However, I am not certain why the fix is working. Before it, the customEditorController was referenced correctly leading to togglePreview() not being defined when it should.

ulferts committed:
adapt cukes to work packages angular show

ulferts committed:
remove redundant cuke

The scenario after the removed one does essentially the same thing with the addition of testing the preview function.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://16364-fullscreen-edit.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://16364-fullscreen-edit.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts committed:
have labels for all the inplace edit fields

The labels are hidden, so they do not disturb. On the other hand, they enable blind users to have the fields properly named for them by a screen reader. It would be better, to have the element to the right of the input element (in the dt) turn into a label. But given the current structure of the directives that would have been way more complicated

ulferts committed:
let gulp watch html templates in all of app

ulferts committed:
fix some cukes & enable others

Some of those still fail because the functionality is actually broken

ulferts committed:
add a root dom element to directive

ulferts committed:
add missing module to directive test

ulferts committed:
fix non terminating digest loop for duration fields

ulferts committed:
adapt cukes to changed implementation

ulferts committed:
enable changes to assignee/responsible to be reflected directly

ulferts committed:
fix showing users right after changing the value

ulferts committed:
remove no longer valid scenario

Time logs can no longer be created when uploading a wp.

ulferts committed:
define placeholders in yml files

ulferts committed:
Merge branch 'dev' into 16364-fullscreen-edit

ulferts committed:
adapt error helper to changed api

ulferts committed:
reenable setting linked resources to null

ulferts committed:
Merge branch 'dev' into 16364-fullscreen-edit

ulferts committed:
remove "save and send" occurence

ulferts committed:
adapt karma unit test to changed api error helper

ulferts committed:
register for later submission upon value change

This is more reliable than the on click handler in place before. With the full edit, the user might not click into the field before starting to edit. He can use tabbing or simply start with the first field that is focused, which is subject.

ulferts committed:
remove merge remains

ulferts committed:
wrap in timeout to trigger digest cycle

ulferts committed:
increase test robustness

ulferts committed:
adapt more cukes to changed implementation

ulferts committed:
allow for issuing a second comment with the same controller

ulferts committed:
improve placeholder robustness

ulferts committed:
correctly initialize comment controller

ulferts committed:
first stab at updating form on type switch

ulferts committed:
increase significance of cuke

ulferts committed:
fix error handling broken in 7577385

ulferts committed:
Merge branch 'dev' into 16364-fullscreen-edit

ulferts committed:
increase spec robustness

element.find(xyz).text will not wait while element.find(xyz, text:text) will.

ulferts committed:
prevent trying to set a non editable field into edit mode

ulferts committed:
Merge branch 'dev' into 16364-fullscreen-edit

ulferts committed:
remove unused variable injection

ulferts committed:
prevent discard on esc when in all edit mode

ulferts committed:
improve cuke robustness by introducing a safeguard

ulferts commented:
There are a lot of code climate warnings. They arise now because @furinvader changed the structure of the code and as such, codeclimate considers the moved code to be new code. I tried to ensure, that no new codeclimate issue arrises.

ulferts commented:
https://community.openproject.org/work_packages/21302

ulferts committed:
Merge branch 'dev' into 16364-fullscreen-edit

ulferts committed:
remove code duplicated from drop down directive

ulferts committed:
adapt dynamic attribute directive to be more in line with other inplace edit directive

furinvader committed:
Merge branch 'dev' into 16364-fullscreen-edit

furinvader committed:
Add disabled attribute to watcher button directive

 Disable the watcher button if the edit all-mode
 is active.
